src/core/geocoding/qgsnominatimgeocoder.cpp

```diff
@@ -0,0 +1,250 @@
+  fields.append( QgsField( QStringLiteral( "place_id" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "class" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "type" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "road" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "village" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "city_district" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "town" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "city" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "state" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "country" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "postcode" ), QVariant::String ) );
+  return fields;
+}
+
+QgsWkbTypes::Type QgsNominatimGeocoder::wkbType() const
+{
+  return QgsWkbTypes::Point;
+}
+
+QList<QgsGeocoderResult> QgsNominatimGeocoder::geocodeString( const QString &string, const QgsGeocoderContext &context, QgsFeedback *feedback ) const
+{
+  QgsRectangle bounds;
+  if ( !context.areaOfInterest().isEmpty() )
+  {
+    QgsGeometry g = context.areaOfInterest();
+    QgsCoordinateTransform ct( context.areaOfInterestCrs(), QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:4326" ) ), context.transformContext() );
+    try
+    {
+      g.transform( ct );
+      bounds = g.boundingBox();
+    }
+    catch ( QgsCsException & )
+    {
+      QgsDebugMsg( "Could not transform geocode bounds to WGS84" );
+    }
+  }
+
+  const QUrl url = requestUrl( string, bounds );
+
+  QgsReadWriteLocker locker( sMutex, QgsReadWriteLocker::Read );
+  auto it = sCachedResults.constFind( url );
+  if ( it != sCachedResults.constEnd() )
+  {
+    return *it;
+  }
+
+  while ( QDateTime::currentMSecsSinceEpoch() - sLastRequestTimstamp < 1000 / mRequestsPerSecond )
+  {
+    QThread::msleep( 50 );
+    if ( feedback && feedback->isCanceled() )
+      return QList<QgsGeocoderResult>();
+  }
+  locker.changeMode( QgsReadWriteLocker::Write );
+  sLastRequestTimstamp = QDateTime::currentMSecsSinceEpoch();
+  locker.unlock();
+
+  QNetworkRequest request( url );
+  QgsSetRequestInitiatorClass( request, QStringLiteral( "QgsNominatimGeocoder" ) );
+
+  QgsBlockingNetworkRequest newReq;
+  const QgsBlockingNetworkRequest::ErrorCode errorCode = newReq.get( request, false, feedback );
+
+  if ( errorCode != QgsBlockingNetworkRequest::NoError )
+  {
+    return QList<QgsGeocoderResult>() << QgsGeocoderResult::errorResult( newReq.errorMessage() );
+  }
+
+  // Parse data
+  QJsonParseError err;
+  QJsonDocument doc = QJsonDocument::fromJson( newReq.reply().content(), &err );
+  if ( doc.isNull() )
+  {
+    return QList<QgsGeocoderResult>() << QgsGeocoderResult::errorResult( err.errorString() );
+  }
+
+  locker.changeMode( QgsReadWriteLocker::Write );
```

Comment:
Drop if above suggestion is applied -- we'd still have the write lock here already